src/program/lwaftr/tests/propbased/genyang.lua

```diff
@@ -0,0 +1,369 @@
+end
+
+-- Like generate_get but for state queries
+function generate_get_state(pid, schema, query)
+   if not query then
+      query, schema = generate_config_xpath_state(schema)
+   end
+   return string.format("./snabb config get-state -s %s %s \"%s\"", schema, pid, query)
+end
+
+-- Used primarily for repeating a set with a value seen before from a get
+function generate_set(pid, schema, query, val)
+   return string.format("./snabb config set -s %s %s \"%s\" \"%s\"",
+                        schema, pid, query, val)
+end
+
+function run_yang(yang_cmd)
+   local f = io.popen(yang_cmd)
+   local result = f:read("*a")
+   f:close()
+   return result
+end
+
+-- choose an element of an array randomly
+local function choose(choices)
+   local idx = math.random(#choices)
+   return choices[idx]
+end
+
+-- choose a natural number (e.g., index or length of array) by
+-- repeating a cointoss
+local function choose_nat()
+   local r = math.random()
+
+   local function flip(next)
+      local r = math.random()
+      if r < 0.5 then
+         return next
+      else
+         return flip(next + 1)
+      end
+   end
+
+   -- evenly weight first two
+   if r < 0.5 then
+      return choose({1, 2})
+   else
+      return flip(3)
+   end
+end
+
+local function random_hex()
+  return string.format("%x", math.random(0, 15))
+end
+
+local function random_hexes()
+   local str = ""
+   for i=1, 4 do
+      str = str .. random_hex()
+   end
+   return str
+end
+
+-- return a random number, preferring boundary values
+local function choose_range(lo, hi)
+   local r = math.random()
+   if r < 0.5 then
+      local mid = math.ceil((hi + lo) / 2)
+      return choose({ lo, lo + 1, mid, mid +  1,  hi - 1, hi })
+   else
+      return math.random(lo, hi)
+   end
+end
+
+local function value_from_type(a_type)
+   local prim = a_type.primitive_type
+
+   if prim == "int8" then
+      return choose_range(-128, 127)
+   elseif prim == "int16" then
+      return choose_range(-32768, 32767)
+   elseif prim == "int32" then
+      return choose_range(-2147483648, 2147483647)
+   elseif prim == "int64" then
+      return choose_range(-9223372036854775809, 9223372036854775807)
+   elseif prim == "uint8" then
+      return choose_range(0, 255)
+   elseif prim == "uint16" then
+      return choose_range(0, 65535)
+   elseif prim == "uint32" then
+      return choose_range(0, 4294967295)
+   elseif prim == "uint64" then
+      return choose_range(0, 18446744073709551615)
+   --elseif prim == "decimal64" then
+   --   local int64 = value_from_type("int64")
+   --   local exp   = math.random(1, 18)
+   --   return int64 * (10 ^ -exp)
+   elseif prim == "boolean" then
+      return choose({ true, false })
+   elseif prim == "ipv4-address" then
+      return math.random(0, 255) .. "." .. math.random(0, 255) .. "." ..
+             math.random(0, 255) .. "." .. math.random(0, 255)
+   elseif prim == "ipv6-address" then
+      local addr = random_hexes()
```

Comment:
This is a fine approach since it's a short string; for longer ones, it makes more sense to build up an array and then use a function that does the whole join in one go.